beets/library.py

```diff
@@ -1064,26 +1064,20 @@ def move(
 
     def destination(
         self,
-        fragment=False,
+        relative_to_libdir=False,
         basedir=None,
-        platform=None,
         path_formats=None,
-        replacements=None,
-    ):
+    ) -> bytes:
         """Return the path in the library directory designated for the
         item (i.e., where the file ought to be).
 
-        fragment makes this method return just the path fragment underneath
-        the root library directory; the path is also returned as Unicode
+        the path is also returned as Unicode
         instead of encoded as a bytestring. basedir can override the library's
         base directory for the destination.
         """
         self._check_db()
-        platform = platform or sys.platform
         basedir = basedir or self._db.directory
```

Comment:
If only it was so simple... 😅 

![image](https://github.com/user-attachments/assets/1d10dd02-28b6-40fd-b50b-c86e888fd16a)

`_check_db` return type says `Database`, while in reality this returns an instance of `Library`, which indeed has `directory` and `path_formats` attributes.

This entire design with an optional `_db` attribute has generally been driving me crazy 🤯.  It's a bit of a hard nut to crack - it's hardly possible to track down how and where `_db` is used because it's just about everywhere in the models code 😁 

I'll see whether I can adjust the type `_check_db` returns to `Library` - I remember when I tried doing it last time I gave up 🤣

Reply:
Something like (untested, I might have mixed up the syntax):

```python
D = TypeVar("D")

class Model(Generic[D]):
    def __init__(self, db: D):
        self._db = db
        ...

    def _check_db(self) -> D:
        ...

class LibModel(Model[D]):
    ...

class Item(Libmodel[Library]):
    ...
```

In Rust, these would "associated types" on a trait. In beets, we do something similar in `dbcore.types`.

Reply:
I see! How would it help to make sure that the `_db` attribute is non-optional?

Reply:
Ignore my question - I initially failed to understand your example properly. Now I see what's going on 😅

Reply:
Though, we know that the `_db` attribute always has the `Library | None` type. Is there any need to have a type variable there in such case?

Reply:
It worked wonderfully: see this PR: https://github.com/beetbox/beets/pull/5545